### DISCUSSION TITLE
Is there any way to make authorization as optional in docs?

### DISCUSSION THREAD

**Discussion by pisces-hui** [fastapi/fastapi#6804] [Questions] [answered]
```
@router.get("/next", summary="Get Next Question", response_model=NextResponse)
def read_item(
    db: Session = Depends(get_db),
    current_user: User = Depends(users.get_current_user),
):
    nextQuestion = crud.get_next_question(db)
    if current_user:
        crud.get_question_is_collect(db, nextQuestion.id, current_user.id)
        nextQuestion.is_collect = True
    return successRes(data=nextQuestion)
```
My requirement is that both guest and user can access this interface, and user can get more data like **nextQuestion.is_collect**.
The question is that I execute in swagger docs, I found the header Authorization was not set even if authorized.
```
curl -X 'GET' \
  'http://127.0.0.1:8000/next' \
  -H 'accept: application/json'
```
What I expected is:
```
curl -X 'POST' \
  'http://127.0.0.1:8000/do_collect_question/1' \
  -H 'accept: application/json' \
  -H 'Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJz……'
```
**(But do not set Authorization if user is not sign in.)**

What should I do?
Any help is appreciated.

**Comment by pisces-hui** [comment, marked as answer]
I got a solution
```
optional_oauth2_scheme = OAuth2PasswordBearer(tokenUrl="auth", auto_error=False)
```

**Comment by tiangolo** [comment]
Yep, that's it! Thanks for reporting back and closing the issue 👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Reply by haoyun** [reply]
However, the generated `openapi.json` is not perfect. It would be better that in this case, an empty `{}` is included in the `security` array, but currently it's not. For example, the given solution produces:

```
        "security": [
          {
            "OAuth2PasswordBearer": []
          }
        ]
```

instead of

```
        "security": [
          {},
          {
            "OAuth2PasswordBearer": []
          }
        ]
```

see for example:

https://github.com/OAI/OpenAPI-Specification/pull/1886
https://github.com/OAI/OpenAPI-Specification/issues/14#issuecomment-297457320

To include `{}` in `security`, we have to use to set `openapi_extra={"security": [{}]}` in the path decorator.

**Reply by haoyun** [reply]
looking at the function

https://github.com/fastapi/fastapi/blob/a96dddb6fdcb90e0c36fde5c4e21542903625e96/fastapi/openapi/utils.py#L77-L91

it seems it's not that possible to hack a `None` security, so that it's automatically added to the schema as `{}`. We have to patch this function to cover that use case.

> To make security optional, an empty security requirement ({}) can be included in the array.
> https://swagger.io/specification/

However, I don't know to obtain the typing information of the `SecurityRequirement`s.

We have to manually declare the `openapi_extra`.
